src/sources/connect/wasm_validators.rs

```diff
@@ -0,0 +1,245 @@
+//! Validation helpers for WASM targets
+//!
+//! These utility methods are mostly used for allowing WASM contexts to validate
+//! connect-enabled schemas. Since this must pass the WASM FFI boundary, they can
+//! only operate on WASM primitives (for now) like strings.
+
+use crate::{
+    error::FederationError,
+    schema::ValidFederationSchema,
+    sources::connect::spec::schema::{
+        ConnectDirectiveArguments, SourceDirectiveArguments, CONNECT_DIRECTIVE_NAME_IN_SPEC,
+        SOURCE_DIRECTIVE_NAME_IN_SPEC,
+    },
+};
+
+#[derive(Default)]
+pub struct Location {
+    pub line: usize,
+    pub column: usize,
+}
+#[derive(Default)]
+pub struct LocationRange {
+    pub start: Location,
+    pub end: Location,
+}
+pub struct ValidationError {
+    pub range: LocationRange,
+    pub reason: String,
+    pub error: FederationError,
+}
+
+/// Validate a subgraph schema with connect directives
+// Note: The error is boxed because it is large and clippy complains
+pub fn validate_connect_enabled_schema(subgraph_schema: &str) -> Result<(), Box<ValidationError>> {
+    // Helper for mapping into the correct error type
+    fn error_with_phony_loc(reason: &str, e: impl Into<FederationError>) -> Box<ValidationError> {
+        Box::new(ValidationError {
+            range: Default::default(),
+            reason: reason.to_string(),
+            error: e.into(),
+        })
+    }
+
+    // First we need to create an actul schema from the provided one
+    let combined_schema = format!(
+        "{}\n{}\n{}",
+        subgraph_schema,
+        constants::TEMP_FEDERATION_DEFINITIONS,
```

Comment:
this shouldn't be necessary, let me check the JS code real quick

Reply:
yep — the subgraphs [at this point](https://github.com/apollographql/federation/blob/main/composition-js/src/compose.ts#L57) are already preprocessed and made valid using the specs (link, federation, connect) so we don't need to edit them before parsing/validating.

Reply:
Sounds good! I'll fix this now